### ISSUE TITLE
Support individual model runs (2 possible implementations)

### ISSUE THREAD

**Issue by dwreeves** [astronomer/astronomer-cosmos#754]
## Context

A common pattern is-- I will deploy a code change, and I will want to either do a full refresh on an existing model, or run a new model so I can see it immediately in production without waiting on the next scheduled DAG run. 

I've been running dbt inside Airflow since prior to Astronomer Cosmos even existing, and this usage pattern has come up a lot in my experience. Right now, I don't think right now that Astronomer Cosmos supports it very well. The recent addition in #623 is much needed for supporting the pattern I am describing, but it does not go far enough.

Right now, in our Astronomer Cosmos deployment, we have a parameterized DAG that allows for this single node invocation, but in order to implement it, I needed to subclass the dbt operator and make everything I wanted into templated fields. I think though that astronomer-cosmos should support this pattern more directly without needing to do that.

## Implementation idea #1 (simpler)

Allow the user to more easily parameterize a DAG with the following additions:

- Make `dbt_cmd` a template field, and allow it to support `str` types (and `if isinstance(dbt_cmd, str): dbt_cmd = [dbt_cmd]`).
- Make `select`, `selector`, etc. all template fields.

This will allow the user to set up a simple DAG that passes in `Param()`s that looks something like this:

```python
with DAG(
    ...,
    params={
        "full_refresh": Param(default=True, ...),
        "dbt_cmd": Param(default="build", ...)
        "select": Param(...)
    },
    render_template_as_native_obj=True
) as dag:
    DbtLocalOperator(
        task_id="dbt_task",
        full_refresh="{{ params.full_refresh }}",
        dbt_cmd=" {{ params.dbt_cmd }}"
        select="{{ params.select }}"
    )
```

The main benefit to this approach compared to the alternative is it does not expand the high level API surface, however it does add functionality to the existing base operator that would be dangerous to ever deprecate. However I do not think the added functionality is hard to support or is even a downside to have; templating these fields could be beneficial to many users.

## Implementation idea #2 (more opinionated, easier to set up)

Set up an "individual run" DAG class for the user, that is pre-parameterized.

```python
from cosmos import DbtParametrizedDag

parametrized_dag = DbtParametrizedDag(
    profile_config=...,
    dag_id="dbt_run_manually",
    schedule_interval=None,
    start_date=...
)
```

There are two advantages to this approach. First, it is easier for the user to set up and would be a supported API. Second and optionally, this can be implemented without impacting the APIs of the dbt base operators, if there is any desire to do that.

**Comment by dwreeves** [comment]
1. An implementation note on the second option, i.e. `DbtParametrizedDag`-- It would likely make sense to use [dynamic task mapping](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/dynamic-task-mapping.html). I think the workflow is you use `DbtLSLocalOperator` (or equivalent) to figure out which nodes get selected; the output is already streamed to xcom via the return value so no changes would be required.

2. Come to think of it, the first option would require a "build" operator, or something else, to work nicely. A big reason why is because you cannot just access `--full-refresh` via the base operator (unless that API decision were to change), meaning a single static operator type cannot handle for users every possible thing they would want to run-- except for the `dbt build` command, which doesn't have an associated operator.

**Comment by joppevos** [comment]
Great description @dwreeves .  I am really interested in this problem!  For me the addition I wrote in #623 has been usable in production. However, I agree that we can improve it even further. I prefer the first solution as abstracting the parameters in their own DAG seems unnecessary and of limited benefit to the end-user. 

> the first option would require a "build" operator, or something else, to work nicely. A big reason why is because you cannot just access --full-refresh via the base operator
Could you clarify this a bit more? We can only run full-refresh on the Seed an RunOperator. Why do we need to access full-refresh from the base operator?

**Comment by dwreeves** [comment]
> Could you clarify this a bit more? We can only run full-refresh on the Seed an RunOperator. Why do we need to access full-refresh from the base operator? The select and full-refresh don't need to be aware of each other right?

The pattern is basically this:

- Create a normal Airflow DAG (not a `cosmos.DbtDag`) with a single dbt operator.
- The DAG is parameterized for the `--select`, `--full-refresh`, and `base_cmd`.
- This operator ideally should be able to run seeds, snapshots, models, tests, etc.

^ Those are the requirements of our little system for manually scheduling dbt nodes.

So how do you achieve this with the current Cosmos API? The answer right now is you need to subclass a dbt operator, and you'd still need to do this even after adding the template fields.

Alternate approaches and why they don't work:

- You _can_ use a dbt run operator, but this means you cannot run other node types like seeds that you may want to trigger.
- You _can_ use the dbt base operator and parametrize the `base_cmd`, but this means no `--full-refresh`.

So basically, the only way to meet the (I think reasonable) requirements of the system is to subclass. Template fields alone don't fulfill the requirements of the system.

Now I'm not necessarily saying the dbt operator should have `--full-refresh`. That is one option though. Another option is to have an operator for `dbt build`.

**Comment by joppevos** [comment]
What do you think of keeping the operations separated? 
More like the example below: ( need to check how to get this working)
```python
with DAG(
    ...,
    params={
        "full_refresh": Param(default=True, ...),
        "select": Param(default="+my_models")
    },
    render_template_as_native_obj=True
) as dag:
    seed = DbtSeedLocalOperator(
        task_id="seed",
        project_dir=CURRENT_DIR,
        profile_config=profile_config,
        full_refresh="{{ params.full_refresh }}",
        select="{{ params.select }}"
    )

    tg = DbtTaskGroup(
        task_id="dbt_task",
        full_refresh="{{ params.full_refresh }}",
        select="{{ params.select }}"
    )
    seed > tg
```

> You can use a dbt run operator, but this means you cannot run other node types like seeds that you may want to trigger.
This could run whatever operators the person has build up within his DAG and given parameters. Then the person does not need to give a parameter for which run operation. 

A user gives the two requested parameters in Airflow Console. The DAG will still be rendered with all nodes, but would only run the selected nodes. If the select contains a seed, then it will run the seed. Otherwise dbt will skip over it. In my case, I only experienced users wanting to run `dbt run` from the console, mostly to backfill. 

Do you feel a strong case for having other dbt commands being trigger from console?

**Comment by dwreeves** [comment]
Sorry, just getting back to this since I am looking to just implement the templating of these fields.

@joppevos A note-- the issue with that code example is that you cannot parametrize the task group. Although Airflow supports dynamically generated tasks, such tasks cannot have inter-dependencies. You could create code that works to dynamically generate all nodes selected via `--select mymodel+`, but the dependencies would not have proper dependencies among each other.

You can have both a seed operator and model run operator and pass to both the same `{{ params.select }}` and that does support; alternatively you can just have two separate DAGs. Or, just do what I am currently doing, and subclass for `dbt build`. Those are all valid options.

I do feel that the most sensible way to have a manual run that triggers on multiple nodes, and multiple node _types_ in a one-shot fashion that doesn't require subclassing is to have an operator that supports `dbt build`. I also do feel this is a good pattern, and I would advocate for it. But I also understand that this is a little more on the niche side, since an operator for `dbt build` would not be used anywhere other than for this specific use case.

**Comment by dwreeves** [comment]
With 1.4.0, I think this is very close to done and this issue can probably be closed.

The `DbtBuildOperator` with `full_refresh` and `select` as templated fields means that setting up a manual, parametrized DAG which runs arbitrary node types does not require any subclassing. (Users who also want to parametrize the command will still need to subclass, but that's fine.)
